docs/source/format/CProtocol.rst

```diff
@@ -0,0 +1,560 @@
+
+.. c:member:: const char* ArrowArray.metadata
+
+   Optional.  A null-terminated, UTF8-encoded string describing
+   the type's metadata.  If the data type is nested, child types are not
+   encoded here but in the :c:member:`ArrowArray.children` arrays.
+
+   MUST be a JSON-compatible mapping of UTF8 strings to UTF8 strings.
+   Whitespace MUST only use the 0x20 character. Example::
+
+      {"key1": "base64-encoded value1", "key2": "base64-encode value2"}
+
+   If omitted, MUST be NULL (not an empty string).
+
+   Consumers MAY choose to ignore metadata.  Even then, they SHOULD keep
+   the metadata string around so as to propagate its information to their
+   own consumers.
+
+.. c:member:: int64_t ArrowArray.flags
+
+   Optional.  A bitfield of flags enriching the type or array description.
+   Its value is computed by OR'ing together the flag values.
+   The following flags are available:
+
+   * ``ARROW_FLAG_NULLABLE``: whether this field is semantically nullable
+     (regardless of whether it actually has null values).
+   * ``ARROW_FLAG_ORDERED``: for dictionary-encoded arrays, whether the
+     ordering of dictionary indices is semantically meaningful.
+
+   If omitted, MUST be 0.
+
+   Consumers MAY choose to ignore some or all of the flags.  Even then,
+   they SHOULD keep this value around so as to propagate its information
+   to their own consumers.
+
+.. c:member:: int64_t ArrowArray.length
+
+   Mandatory.  The logical length of the array (i.e. its number of items).
+
+.. c:member:: int64_t ArrowArray.null_count
+
+   Mandatory.  The number of null items in the array.  MAY be -1 if not
+   yet computed.
+
+.. c:member:: int64_t ArrowArray.offset
+
+   Mandatory.  The logical offset inside the array (i.e. the number of items
+   from the physical start of the buffers).  MUST be 0 or positive.
+
+   Producers MAY specify that they will only produce 0-offset arrays to
+   ease implementation of consumer code.
+   Consumers MAY decide not to support non-0-offset arrays, but they
+   should document this limitation.
+
+.. c:member:: int64_t ArrowArray.n_buffers
+
+   Mandatory.  The number of physical buffers backing this array.  The
+   number of buffers is a function of the data type, as described in the
+   :ref:`Columnar format specification <format_columnar>`.
+
+   Buffers of children arrays are not included.
+
+.. c:member:: const void** ArrowArray.buffers
+
+   Mandatory.  A C array of pointers to the start of each physical buffer
+   backing this array.  Each `void*` pointer is the physical start of
+   a contiguous buffer.  There must be :c:member:`ArrowArray.n_buffers` pointers.
+
+   The producer MUST ensure that each contiguous buffer is large enough to
+   represent `length + offset` values encoded according to the
+   :ref:`Columnar format specification <format_columnar>`.
+
+   The pointer to the null bitmap buffer, if the data type specifies one,
+   MAY be NULL only if :c:member:`ArrowArray.null_count` is 0.
```

Comment:
Is it possible that the null_count is 0, and the bitmap buffer is non-empty? 
I want to know if the null bitmap buffer is created lazily, eagerly, or implementation dependent?

Reply:
It is possible. In some situations you don't know that there are zero nulls upfront, and you have a buffer.

Reply:
I see. Thanks.